### PR TITLE
Phase 1 x cleanup backend

### PR DESCRIPTION
**Pull request by KonstZiv**
# Phase 1.X — Cleanup bundle (backend)

Phase 1.X cleanup bundle resolving 7 deferred items з Phase 1 POST-MERGE-NOTES §5 та AI ReviewBot Round 1+2 deferred. Single PR з 4 atomic commits, no investigation phase (light cleanup process per CLEANUP-PLAN).

## Items resolved

| Item | Description | Origin |
|---|---|---|
| 5 | S3Client encapsulation — public `get_object_url(key)` method, removed direct access to `_endpoint_url` / `_bucket` from route handlers + DRY internal callers (`upload_file`, `upload_smart`) | AI ReviewBot [2] |
| 8 | Filename sanitization — `sanitize_s3_key(filename)` helper using `PurePosixPath.name` для path traversal prevention; applied у 2 user-derived callsites (`create_document`, `get_upload_url`); skip `confirm_upload` (tenant-prefix-validated body.key) | AI ReviewBot [16] |
| 3 | Amendment 31 — rename `NodeWithMaterialsResponse` → `NodeWithDocumentsResponse` (Pydantic class identifier alignment з KD-ui.3 `authored_documents` JSON contract); 5 callsites across 2 files | Amendment 31 |
| 7 | Doc accuracy fixes у `tests/integration/conftest.py` + `tests/unit/test_api/test_reconciliation_routes.py` — tautology cleanup post-Phase-1 sed pass | AI ReviewBot [10][11][12][13] |
| 9 | DRY refactor — extract `_MARKER_REGEX` + `_assert_marker_recent` до `tests/_helpers/kd3_marker.py` (shared module); replace 14 callsites across 2 test files (5 + 9) | AI ReviewBot [17] |
| 4 | `make psql` target — DB probe convenience using existing `up` target env-fallback convention (`$${POSTGRES_USER:-course_supporter}`) | CHECKPOINT 2 vision-side observation |

## Items dropped

**Item 6 (`router=router` "dead code" у `arq_ingest_material`)** — DROPPED 2026-05-07 as false-positive per own probe.

Receivers verification:
- `create_heavy_steps` (factory.py:74-79) — uses router conditionally.
- `create_vd_pipeline` (factory.py:113-115) — uses router у VDPipeline construction.
- `processor.process_raw` — polymorphic across 5 implementations: `TextProcessor` explicit-ignores per protocol conformance docstring; `VideoProcessor` / `WhisperVideoProcessor` / `LongVideoProcessor` / `PresentationProcessor` / `WebProcessor` all use router для Gemini Vision LLM routing.

Видалення зламало б video + presentation production pipelines. AI ReviewBot [4] = false-positive (limited diff context misclassified polymorphic callsite).

## Items deferred

**Item 3 hotfix-10 (Amendment 35 sub-class 3b — `_collect_ready_documents` runtime-broken block)** — out of scope, separate Phase 1.X candidate per `current-state.md`. Conditional resolution: standalone hotfix if feature activates pre-Phase 5, else deletion alongside `arq_ingest_material` orchestration rewrite у Phase 5.

## Commits

| # | SHA | Subject |
|---|---|---|
| 1 | e3c8f04 | `phase(1.X.backend-1): S3Client encapsulation + filename sanitization` |
| 2 | 93d06a2 | `phase(1.X.backend-2): rename NodeWithMaterialsResponse → NodeWithDocumentsResponse (Amendment 31)` |
| 3 | 6b617ec | `phase(1.X.tests): doc accuracy fixes + DRY refactor _MARKER_REGEX helpers` |
| 4 | 7107059 | `phase(1.X.tooling): add make psql target for DB probe convenience` |

Aggregate: 13 files changed (3 new + 10 modified). 4 commits ahead `origin/dev` 8304870.

## Acceptance evidence

Per-commit quality gates (all green):

- **9 pre-commit hooks** active per commit (varying applicability — Makefile/test commits skip ruff/mypy where no Python touched).
- **mypy --strict** на всіх modified production files: 0 issues. Recursive Pydantic forward-reference у `children: list[NodeWithDocumentsResponse]` resolved cleanly without `model_rebuild()`.
- **ruff format + check**: clean across all touched files.
- **Narrow test sweep** per Amendment 36:
  - backend-1: 59 passed / 0 failed (`tests/unit/test_s3_client.py` + 2 route test files).
  - backend-2: 56 passed / 7 skipped requires_db (`tests/unit/test_api/test_nodes.py` + cascade tests).
  - tests: 58 passed / 0 failed (`tests/unit/test_cascade_delete_service.py` + `tests/storage/test_cascade_invalidation.py`).
  - tooling: manual gate (Makefile change).
- **Verification taxonomy** per Amendment 37: changeset = (a) true fix + new coverage; 0 of (b)/(c)/(d).

## Cross-references

- `phase-1-X/state.md` — full ratification chain + per-sub-area evidence.
- `vision.md` v0.20.2 §3 KD3 (cascade soft-delete) + KD9 (content_hash naming) + Amendment 31 (NodeWithDocumentsResponse rename) + Amendment 35 sub-classifications + Amendment 37 verification taxonomy.
- POST-MERGE-NOTES §5 (Phase 1 deferred debt list) — items 4, 5, 7, 8, 9, 17 closed by this PR; Item 6 closed as false-positive; Item 3/16 (Amendment 31) closed.
- `phase-1-X-CLEANUP-PLAN.md` — bundle scope authoritative reference.

## Process discipline lessons

**Rule #25 candidate (vision-side broad-claim verification) — 5th validation instance:** implementer stale memory anchor про phase ownership ("Phase 5 territory per Amendment 35 sub-class 3a" claim) caught by vision-side STOP-escalate during pre-flight review. Vision.md v0.20.2 explicit Phase 1.1 sub-phase definition — `FingerprintService` collapse IS Phase 1.1, не Phase 5. Fix would have introduced incorrect future-state into previously-correct comments. Rule approaching promotion criteria (validated through Phase 1.X bundle).

**Pattern validated:** "light cleanup PR" template (mini pre-flight ~5 точок per item, atomic commit per Rule #13, push per Rule #6, PR creation HOLD до bundle completion) successful — 4 commits, 0 mid-bundle escalation cycles, 1 vision-side STOP-escalate caught false-positive scope expansion at pre-flight stage (not post-commit).

## OpenAPI impact

Component name shift `NodeWithMaterialsResponse` → `NodeWithDocumentsResponse` у `/openapi.json`. JSON wire shape unchanged (field names + types + `validation_alias="documents"` bridge preserved). UI codebase prepared per KD-ui.8 trail.

## Review request

AI ReviewBot — primary review surface per Phase 1 cadence. Manual review focus areas:
- `S3Client.get_object_url` URL formula correctness (path-style, trailing slash handling).
- `sanitize_s3_key` edge cases (empty / dot-only / Windows-style backslash inputs).
- DRY refactor у `tests/_helpers/kd3_marker.py` — verify zero behavior divergence vs original inline definitions.

### CODE REVIEW

File: tests/_helpers/kd3_marker.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+        int(hour),
+        int(minute),
+        int(second),
+        tzinfo=UTC,
+    )
+
+
+def assert_marker_recent(value: str) -> None:
+    """Assert ``value`` matches the KD3 marker contract with a recent
+    timestamp (within ``TIMESTAMP_TOLERANCE_SECONDS`` of now).
```
[the remark this answers is not in the document]
Comment (KonstZiv) [reply tests/_helpers/kd3_marker.py:46]: 5-second tolerance is intentional — accommodates marker-creation-to-assertion gap у normal test execution. Pattern pre-existed identically у both source files pre-DRY-refactor; no flakiness observed across Phase 1 + cleanup chain CI runs. Alternative (mock datetime.now() у production code) widens scope beyond Phase 1.X cleanup boundary. If flakiness materializes у future CI runs — addressing as standalone hotfix.
